pkg/proxy/config/api_test.go

```diff
@@ -190,25 +164,59 @@ func TestInitialSync(t *testing.T) {
 		ObjectMeta: metav1.ObjectMeta{Namespace: "testnamespace", Name: "bar"},
 	}
 
-	var wg sync.WaitGroup
-	// Wait for both services and endpoints handler.
-	wg.Add(2)
+	expectedSvcState := map[types.NamespacedName]*v1.Service{
+		{Name: svc1.Name, Namespace: svc1.Namespace}: svc1,
+		{Name: svc2.Name, Namespace: svc2.Namespace}: svc2,
+	}
+	expectedEpsState := map[types.NamespacedName]*v1.Endpoints{
+		{Name: eps1.Name, Namespace: eps1.Namespace}: eps1,
+		{Name: eps2.Name, Namespace: eps2.Namespace}: eps2,
+	}
 
 	// Setup fake api client.
 	client := fake.NewSimpleClientset(svc1, svc2, eps2, eps1)
 	sharedInformers := informers.NewSharedInformerFactory(client, 0)
 
 	svcConfig := NewServiceConfig(sharedInformers.Core().V1().Services(), 0)
-	epsConfig := NewEndpointsConfig(sharedInformers.Core().V1().Endpoints(), 0)
-	svcHandler := newSvcHandler(t, []*v1.Service{svc2, svc1}, wg.Done)
+	svcHandler := &ServiceHandlerMock{
+		state:   make(map[types.NamespacedName]*v1.Service),
+		process: func(services []*v1.Service) {},
+	}
 	svcConfig.RegisterEventHandler(svcHandler)
-	epsHandler := newEpsHandler(t, []*v1.Endpoints{eps2, eps1}, wg.Done)
+
+	epsConfig := NewEndpointsConfig(sharedInformers.Core().V1().Endpoints(), 0)
+	epsHandler := &EndpointsHandlerMock{
+		state:   make(map[types.NamespacedName]*v1.Endpoints),
+		process: func(endpoints []*v1.Endpoints) {},
+	}
 	epsConfig.RegisterEventHandler(epsHandler)
 
 	stopCh := make(chan struct{})
 	defer close(stopCh)
-	go sharedInformers.Start(stopCh)
+	sharedInformers.Start(stopCh)
 	go svcConfig.Run(stopCh)
 	go epsConfig.Run(stopCh)
-	wg.Wait()
+
+	err := wait.PollImmediate(time.Millisecond*10, wait.ForeverTestTimeout, func() (done bool, err error) {
+		svcHandler.lock.Lock()
+		defer svcHandler.lock.Unlock()
+		if reflect.DeepEqual(svcHandler.state, expectedSvcState) {
+			return true, nil
```

Comment:
Do you have any objections to returning the more verbose error with details on the current versus expected not being equal? Is the more concise error suggesting that a timeout would be the only reason this would fail?

Reply:
> Do you have any objections to returning the more verbose error with details on the current versus expected not being equal

I think we could add some logging if what we got was different from the expected, but it might not be desired to return an error immediately as the poll would be interrupted.

> Is the more concise error suggesting that a timeout would be the only reason this would fail?

yes